File: apps/desktop/frontend/src/core/stream-deck-controller/commands.ts
export enum StreamDeckAction {
  // Track selection (send track number as payload)
  SelectTrack = 'SELECT_TRACK',
  Undo = 'UNDO',

  // Active track actions
  PrevTrack = 'PREV_TRACK',
  NextTrack = 'NEXT_TRACK',
  MuteActive = 'MUTE_ACTIVE',
  ClearActive = 'CLEAR_ACTIVE',
  ToggleMetronome = 'TOGGLE_METRONOME',
  MergeWithBelow = 'MERGE_WITH_BELOW',
  MergeWithAbove = 'MERGE_WITH_ABOVE',

  // Transport
  PlayPause = 'PLAY_PAUSE',
  Record = 'RECORD',
  Overdub = 'OVERDUB',
  Stop = 'STOP',
  Shift = 'SHIFT',
}

export interface StreamDeckCommand {
  action: StreamDeckAction
  payload?: unknown
}

export const STREAM_DECK_LAYOUT: StreamDeckCommand[][] = [
  [
    { action: StreamDeckAction.SelectTrack, payload: 1 },
    { action: StreamDeckAction.SelectTrack, payload: 2 },
    { action: StreamDeckAction.SelectTrack, payload: 3 },
    { action: StreamDeckAction.SelectTrack, payload: 4 },
    { action: StreamDeckAction.Undo },
  ],
  [
    { action: StreamDeckAction.PrevTrack },
    { action: StreamDeckAction.NextTrack },
    { action: StreamDeckAction.MuteActive },
    { action: StreamDeckAction.ClearActive },
    { action: StreamDeckAction.ToggleMetronome },
  ],
  [
    { action: StreamDeckAction.PlayPause },
    { action: StreamDeckAction.Record },
    { action: StreamDeckAction.Overdub },
    { action: StreamDeckAction.Stop },
    { action: StreamDeckAction.Shift },
  ],
]

File: apps/desktop/internal/streamdeck/commands.go
package streamdeck

type ActionType string

const (
	// Row 1 — Direct track selection
	ActionSelectTrack1 ActionType = "SELECT_TRACK_1"
	ActionSelectTrack2 ActionType = "SELECT_TRACK_2"
	ActionSelectTrack3 ActionType = "SELECT_TRACK_3"
	ActionSelectTrack4 ActionType = "SELECT_TRACK_4"
	ActionUndo         ActionType = "UNDO"

	// Row 2 — Active track actions
	ActionPrevTrack       ActionType = "PREV_TRACK"
	ActionNextTrack       ActionType = "NEXT_TRACK"
	ActionMuteActive      ActionType = "MUTE_ACTIVE"
	ActionClearActive     ActionType = "CLEAR_ACTIVE"
	ActionToggleMetronome ActionType = "TOGGLE_METRONOME"

	// Row 3 — Transport
	ActionPlayPause ActionType = "PLAY_PAUSE"
	ActionRecord    ActionType = "RECORD"
	ActionOverdub   ActionType = "OVERDUB"
	ActionStop      ActionType = "STOP"
	ActionShift     ActionType = "SHIFT"
)

type Command struct {
	Action  ActionType `json:"action"`
	Payload any        `json:"payload,omitempty"`
}
